### ISSUE TITLE
opEquals fails to compile with separate compilation

### ISSUE THREAD

**Issue by atilaneves** [pbackus/sumtype#15]
As of master on commit 7c8ad0c, this fails to compile with `dub build --build-mode=singleFile`:

```d
// app.d
void main() {
    import node: Node;
    auto n0 = Node();
    auto n1 = Node();
    pragma(msg, typeof(n0 == n1));
}
```

```d
// node.d
import sumtype;

alias Node = SumType!(Struct, String);

struct Struct {
    string spelling;
    Node[] nodes;
}


struct String {
    string value;
}
```

**Comment by pbackus** [comment]
Not sure yet what's going on here, but changing the definition of `Node` to the following causes `dmd` to segfault:

```d
alias Node = SumType!(
    Tuple!(
        string, "spelling",
        This[], "nodes"
    ),
    String
);
```

**Comment by pbackus** [comment]
The issue here is that `isEqualityComparable!Struct` is true when compiling together, but false when compiling separately. Testing `__traits(compiles, Struct.init == Struct.init)` gives the same result, so this probably isn't a Phobos issue.

Unconditionally generating `SumType.opEquals` fixes this issue, but causes compilation to fail when instantiating `SumType` with a struct that has a `@disable`-ed `opEquals` overload, so it's not a clean solution. Removing `SumType.opEquals` entirely also works, albeit at the cost of breaking #7 again.

I'm at a bit of a loss as to how to proceed here. Between this and [dlang issue 19407](https://issues.dlang.org/show_bug.cgi?id=19407), it seems as though compile-time reflection simply can't be relied upon when using separate compilation. Given that `SumType` has no choice but to rely on compile-time reflection, that's a pretty serious problem.

**Comment by pbackus** [comment]
Submitted a reduced example as [issue 19417](https://issues.dlang.org/show_bug.cgi?id=19417).

**Comment by pbackus** [comment]
This specific issue was fixed in commit 412c0e5, which was released in `sumtype` version 0.8.0. The remaining issues with `opEquals` are documented as known bugs, and require a fix for [dlang issue 19458](https://issues.dlang.org/show_bug.cgi?id=19458) before they can be resolved.
